rust-src/concordium_base/src/id/sigma_protocols/com_ineq.rs

```diff
@@ -0,0 +1,163 @@
+
+pub fn verify_com_ineq<C: Curve>(
+    com_key: &CommitmentKey<C>,
+    c: &Commitment<C>,
+    pub_value: C::Scalar,
+    proof: &Witness<C>,
+) -> bool {
+    let Witness {
+        com_mult_witness,
+        aux_com,
+    } = proof;
+    let mut transcript = RandomOracle::domain(b"NonEqualityProof");
+    transcript.append_message(b"g", &com_key.g);
+    transcript.append_message(b"h", &com_key.h);
+    transcript.append_message(b"public commitment", &c);
+    transcript.append_message(b"public value", &pub_value);
+    let mut minus_pub_value = pub_value;
+    minus_pub_value.negate();
+    let g_minus_pub_value = com_key.g.mul_by_scalar(&minus_pub_value);
+
+    let cmm_1: C = c.plus_point(&g_minus_pub_value);
+    let cmm_3 = Commitment(com_key.g);
+
+    let com_mult = ComMult {
+        cmms:    [Commitment(cmm_1), *aux_com, cmm_3],
+        cmm_key: *com_key,
+    };
+    sigma_verify(&mut transcript, &com_mult, com_mult_witness)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use ff::PrimeField;
+    use pairing::bls12_381::{Fr, G1};
+
+    #[test]
+    fn test_com_ineq_correctness() {
+        let mut csprng = rand::thread_rng();
+        let com_key: CommitmentKey<_> = CommitmentKey::<G1>::generate(&mut csprng);
+
+        let value = Value::<G1>::new(Fr::from_str("20000102").unwrap());
+        let pub_value = Fr::from_str("20000103").unwrap();
+
+        let (cmm_1, value_tilde) = com_key.commit(&value, &mut csprng);
+
+        let proof = prove_com_ineq(&com_key, &value, value_tilde, pub_value, &mut csprng)
+            .expect("Proving should succeed.");
+        assert!(
+            verify_com_ineq(&com_key, &cmm_1, pub_value, &proof),
+            "Incorrect inequality proof."
+        );
+    }
+
+    #[test]
+    fn test_com_ineq_soundness() {
+        let mut csprng = rand::thread_rng();
+        let com_key: CommitmentKey<_> = CommitmentKey::<G1>::generate(&mut csprng);
+
+        // Generate a valid proof
+        let value = Value::<G1>::new(Fr::from_str("20000102").unwrap());
+        let pub_value = Fr::from_str("20000103").unwrap();
+        let (cmm_1, value_tilde) = com_key.commit(&value, &mut csprng);
+        let proof = prove_com_ineq(&com_key, &value, value_tilde, pub_value, &mut csprng)
+            .expect("Proving should succeed.");
+
+        // Make the proof invalid by changing parameters
+        let wrong_com_key = CommitmentKey::<G1>::generate(&mut csprng);
+        let wrong_pub_value = Fr::from_str("20000102").unwrap();
+        let wrong_proof = Witness {
+            aux_com:          Commitment(G1::generate(&mut csprng)),
+            com_mult_witness: proof.com_mult_witness.clone(),
+        };
+
+        // Verify failure for invalid parameters
+        assert!(!verify_com_ineq(&wrong_com_key, &cmm_1, pub_value, &proof));
+        assert!(!verify_com_ineq(&com_key, &cmm_1, wrong_pub_value, &proof));
+        assert!(!verify_com_ineq(&com_key, &cmm_1, pub_value, &wrong_proof));
+    }
+}
```

Comment:
Maybe add a test calling prove for values that are equal.

Reply:
Done!